File: flutter_app/lib/widgets/live_list.dart
import 'package:flutter/material.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../theme/app_theme.dart';
import '../services/data_service.dart';
import 'glass.dart';

/// Renders real rows from a Supabase [table] and live-updates via realtime.
class LiveList extends StatefulWidget {
  const LiveList({super.key, required this.table, required this.icon, this.limit = 50});
  final String table;
  final IconData icon;
  final int limit;

  @override
  State<LiveList> createState() => _LiveListState();
}

class _LiveListState extends State<LiveList> {
  final _svc = DataService();
  RealtimeChannel? _channel;
  List<Map<String, dynamic>>? _rows;
  String? _error;
  bool _live = false;

  @override
  void initState() {
    super.initState();
    _load();
    _channel = _svc.subscribe(widget.table, () {
      _live = true;
      _load();
    });
  }

  @override
  void dispose() {
    _svc.unsubscribe(_channel);
    super.dispose();
  }

  Future<void> _load() async {
    try {
      final rows = await _svc.fetch(widget.table, limit: widget.limit);
      if (!mounted) return;
      setState(() {
        _rows = rows;
        _error = null;
      });
    } catch (e) {
      if (!mounted) return;
      setState(() => _error = e.toString());
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_error != null) return _state(Icons.lock_outline, 'Restricted', 'This table isn\'t readable for your role yet.');
    if (_rows == null) {
      return const Center(child: CircularProgressIndicator(color: AppColors.primary));
    }
    if (_rows!.isEmpty) return _state(widget.icon, 'No records', 'Nothing here yet — new rows appear instantly.');

    return RefreshIndicator(
      color: AppColors.primary,
      onRefresh: _load,
      child: ListView.separated(
        padding: const EdgeInsets.fromLTRB(16, 8, 16, 120),
        itemCount: _rows!.length + 1,
        separatorBuilder: (_, i) => i == 0 ? const SizedBox.shrink() : const SizedBox(height: 8),
        itemBuilder: (context, i) {
          if (i == 0) return _liveHeader();
          final row = _rows![i - 1];
          return _rowCard(row);
        },
      ),
    );
  }

  Widget _liveHeader() => Padding(
        padding: const EdgeInsets.only(bottom: 10),
        child: Row(children: [
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 5),
            decoration: BoxDecoration(
              color: AppColors.primary.withValues(alpha: 0.1),
              borderRadius: BorderRadius.circular(AppRadius.pill),
            ),
            child: Row(mainAxisSize: MainAxisSize.min, children: [
              Container(width: 7, height: 7, decoration: const BoxDecoration(color: AppColors.primary, shape: BoxShape.circle)),
              const SizedBox(width: 6),
              Text(_live ? 'LIVE · updated' : 'LIVE',
                  style: const TextStyle(fontSize: 10, fontWeight: FontWeight.w900, color: AppColors.primary)),
            ]),
          ),
          const Spacer(),
          Text('${_rows!.length} records', style: TextStyle(fontSize: 11, fontWeight: FontWeight.w700, color: AppColors.muted)),
        ]),
      );

  Widget _rowCard(Map<String, dynamic> row) {
    final title = DataService.title(row);
    final sub = DataService.subtitle(row);
    return GlassCard(
      padding: const EdgeInsets.all(12),
      radius: AppRadius.md,
      opacity: 0.55,
      child: Row(children: [
        Container(
          width: 40, height: 40,
          decoration: BoxDecoration(color: AppColors.primary.withValues(alpha: 0.1), borderRadius: BorderRadius.circular(10)),
          child: Icon(widget.icon, size: 19, color: AppColors.primary),
        ),
        const SizedBox(width: 12),
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(title, maxLines: 1, overflow: TextOverflow.ellipsis,
                  style: TextStyle(fontSize: 14, fontWeight: FontWeight.w700, color: AppColors.ink)),
              if (sub.isNotEmpty)
                Text(sub, maxLines: 1, overflow: TextOverflow.ellipsis,
                    style: TextStyle(fontSize: 11, color: AppColors.muted)),
            ],
          ),
        ),
      ]),
    );
  }

  Widget _state(IconData icon, String title, String msg) => Center(
        child: Padding(
          padding: const EdgeInsets.all(28),
          child: GlassCard(
            padding: const EdgeInsets.all(26),
            child: Column(mainAxisSize: MainAxisSize.min, children: [
              Icon(icon, size: 40, color: AppColors.primary),
              const SizedBox(height: 14),
              Text(title, style: TextStyle(fontSize: 17, fontWeight: FontWeight.w900, color: AppColors.ink)),
              const SizedBox(height: 6),
              Text(msg, textAlign: TextAlign.center, style: TextStyle(fontSize: 13, color: AppColors.muted)),
            ]),
          ),
        ),
      );
}

File: android_native/app/src/main/java/com/ezbillify/ezworkspace/ui/components/LiveList.kt
package com.ezbillify.ezworkspace.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.ezbillify.ezworkspace.services.DataService
import com.ezbillify.ezworkspace.theme.AppColors
import com.ezbillify.ezworkspace.theme.InterFontFamily
import com.ezbillify.ezworkspace.theme.neumorphicRaised
import com.ezbillify.ezworkspace.theme.neumorphicSoft
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonObject

/** Renders real rows from a Supabase [table] and live-updates via realtime. Mirrors the Flutter build's LiveList. */
@Composable
fun LiveList(table: String, icon: ImageVector, limit: Long = 50) {
    val svc = remember { DataService() }
    val scope = rememberCoroutineScope()
    var rows by remember { mutableStateOf<List<JsonObject>?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var live by remember { mutableStateOf(false) }
    var refreshing by remember { mutableStateOf(false) }

    suspend fun load() {
        try {
            rows = svc.fetch(table, limit)
            error = null
        } catch (e: Exception) {
            error = e.message ?: "Unable to load this table."
        }
    }

    DisposableEffect(table) {
        val ch = svc.subscribe(table, scope) {
            live = true
            scope.launch { load() }
        }
        scope.launch { load() }
        onDispose { scope.launch { svc.unsubscribe(ch) } }
    }

    when {
        error != null -> EmptyState(Icons.Outlined.Lock, "Restricted", "This table isn't readable for your role yet.")
        rows == null -> Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
            CircularProgressIndicator(color = AppColors.primary)
        }
        rows!!.isEmpty() -> EmptyState(icon, "No records", "Nothing here yet — new rows appear instantly.")
        else -> PullToRefreshBox(
            isRefreshing = refreshing,
            onRefresh = {
                scope.launch {
                    refreshing = true
                    load()
                    refreshing = false
                }
            },
        ) {
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(start = 16.dp, end = 16.dp, top = 8.dp, bottom = 120.dp),
            ) {
                item { LiveHeader(live, rows!!.size) }
                items(rows!!) { row -> Spacer(Modifier.size(8.dp)); RowCard(row, icon) }
            }
        }
    }
}

@Composable
private fun LiveHeader(live: Boolean, count: Int) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(bottom = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Row(
            modifier = Modifier
                .background(AppColors.primary.copy(alpha = 0.1f), RoundedCornerShape(999.dp))
                .padding(horizontal = 10.dp, vertical = 5.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(Modifier.size(7.dp).background(AppColors.primary, CircleShape))
            Spacer(Modifier.width(6.dp))
            Text(
                text = if (live) "LIVE · updated" else "LIVE",
                fontFamily = InterFontFamily, fontWeight = FontWeight.Black, fontSize = 10.sp,
                color = AppColors.primary,
            )
        }
        Spacer(Modifier.weight(1f))
        Text("$count records", fontFamily = InterFontFamily, fontWeight = FontWeight.Bold, fontSize = 11.sp, color = AppColors.muted)
    }
}

@Composable
private fun RowCard(row: JsonObject, icon: ImageVector) {
    val title = DataService.title(row)
    val sub = DataService.subtitle(row)
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .neumorphicSoft(cornerRadius = 14.dp)
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier.size(40.dp).background(AppColors.primary.copy(alpha = 0.1f), RoundedCornerShape(10.dp)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, contentDescription = null, tint = AppColors.primary)
        }
        Spacer(Modifier.width(12.dp))
        androidx.compose.foundation.layout.Column {
            Text(title, maxLines = 1, fontFamily = InterFontFamily, fontWeight = FontWeight.Bold, fontSize = 14.sp, color = AppColors.ink)
            if (sub.isNotEmpty()) {
                Text(sub, maxLines = 1, fontFamily = InterFontFamily, fontSize = 11.sp, color = AppColors.muted)
            }
        }
    }
}

@Composable
private fun EmptyState(icon: ImageVector, title: String, msg: String) {
    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        androidx.compose.foundation.layout.Column(
            modifier = Modifier.padding(28.dp).neumorphicRaised(cornerRadius = 28.dp).padding(26.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Icon(icon, contentDescription = null, tint = AppColors.primary, modifier = Modifier.size(40.dp))
            Spacer(Modifier.size(14.dp))
            Text(title, fontFamily = InterFontFamily, fontWeight = FontWeight.Black, fontSize = 17.sp, color = AppColors.ink)
            Spacer(Modifier.size(6.dp))
            Text(msg, fontFamily = InterFontFamily, fontSize = 13.sp, color = AppColors.muted)
        }
    }
}
